### ISSUE TITLE
Revise section "Exceptions" and "Exceptions and transactions"

### ISSUE THREAD

**Issue by jugmac00** [zopefoundation/Zope#587]
I am stuck a bit revising above mentioned sections in 
https://zope.readthedocs.io/en/latest/zdgbook/ObjectPublishing.html#exceptions
as it seems there have been some massive changes.

What I have found out already:
- With the new WSGI setup there is no more exception type to http resonse code matching, as the imported status_reasons / generated status_codes are no more used
- all unhandled execeptions result in 
```
Internal Server Error
The server encountered an unexpected internal server error
(generated by waitress)
```
- except you implement a standard_error_message - which has to be a callable - unlike the comment suggests
https://github.com/zopefoundation/Zope/blob/2139f2ceeb49380d3dd58388236e0461446096ab/src/ZPublisher/WSGIPublisher.py#L115-L119

better comment:
https://github.com/zopefoundation/Zope/blob/297154fba9f5b7ce71a82b41a70a9ae34ab0fa3e/src/OFS/browser/__init__.py#L25-L27

Any suggestion what to do with chapter https://zope.readthedocs.io/en/latest/zdgbook/ObjectPublishing.html#exceptions ?

Removing the obsolete content there is not much left. I could mention that you need to setup a standard_error_message.

Also, any suggestions about paragraphs 3 and 4 of https://zope.readthedocs.io/en/latest/zdgbook/ObjectPublishing.html#exceptions-and-transactions ?

Except for the "DTML only" reference(opposed to page template, and other callables, too), they seem quit okish.

P.S.: It seems a bit odd that in SimpleItem there is an attempt to acquire **standard_error_message** but none comes with a standard Zope installation or gets autogenerated.

What is the story behind a fresh Zope installation with all unhandled exceptions show this standard wsgi error page?

Also, I am not sure what "Exception views" are and how to make use of them. Anybody can give me some input?

**Comment by dataflake** [comment]
> I am stuck a bit revising above mentioned sections in
> https://zope.readthedocs.io/en/latest/zdgbook/ObjectPublishing.html#exceptions
> as it seems there have been some massive changes.

Yes, and that has to do with the move to WSGI.

> * With the new WSGI setup there is no more exception type to http resonse code matching, as the imported status_reasons / generated status_codes are no more used
> * all unhandled execeptions result in
> 
> ```
> Internal Server Error
> The server encountered an unexpected internal server error
> (generated by waitress)
> ```

All unhandled exceptions propagate up the WSGI stack and that's the default error message waitress provides. Unless you pass the flag "expose_tracebacks" in the WSGI configuration .ini file, then you will see full tracebacks.

> * except you implement a standard_error_message - which has to be a callable - unlike the comment suggests
>   https://github.com/zopefoundation/Zope/blob/2139f2ceeb49380d3dd58388236e0461446096ab/src/ZPublisher/WSGIPublisher.py#L115-L119
> 
> better comment:
> https://github.com/zopefoundation/Zope/blob/297154fba9f5b7ce71a82b41a70a9ae34ab0fa3e/src/OFS/browser/__init__.py#L25-L27

The comment makes no suggestion about callable or not. The code uses queryMultiAdapter to look up a view class. A view instance has a `__call__` method that does rendering. That's "view classes 101" so to speak. That second comment is in the actual view class that is looked up where the first comment is. Only that view class has to know what the old standard_error_message is and how to handle that. The Publisher does not. It just knows about view classes.

> Removing the obsolete content there is not much left. I could mention that you need to setup a standard_error_message.

You need to define a goal before you declare a need to set up something. You don't need to set up anything unless you have a specific goal, like "I want to influence what people see when they get an error". The code in question was put in by me to support migrated instances that have a standard_error_message and want to continue using it. Zope 4, unlike Zope 2, does not generate such a DTML object automatically anymore. It is too unspecific as a catch-all.


> P.S.: It seems a bit odd that in SimpleItem there is an attempt to acquire **standard_error_message** but none comes with a standard Zope installation or gets autogenerated.

That's old existing code, but it is still being used. By the exact same view class that I wrote.

> What is the story behind a fresh Zope installation with all unhandled exceptions show this standard wsgi error page?

Just standard behavior with WSGI. It is up to the application programmer to set up a different solution, like registering exception view or - with the help of the view class I wrote - add a standard_error_message.

> Also, I am not sure what "Exception views" are and how to make use of them. Anybody can give me some input?

Even if you don't find it in our own documentation, Google is your friend. I typed in "Zope exception view" and the second hit was an article by Michael Howitz that explains what such a view is and has great code examples to create and register one: https://blog.gocept.com/2017/10/24/zope4-errorhandling/. Matter of fact (sounding like a broken record again) the standard_error_message view class I wrote is based on his ideas and suggestions.

**Comment by d-maurer** [comment]
Jürgen Gmach wrote at 2019-5-3 15:52 -0700:
> ...
>What I have found out already:
>- With the new WSGI setup there is no more exception type to http resonse code matching, as the imported status_reasons / generated status_codes are no more used
>- all unhandled execeptions result in

Fortunately, some standard exceptions are still mapped (as it must be
to be HTTP compatible): e.g. "NotFound", "Redirect", "Unauthorized", ...

**Comment by jugmac00** [comment]
Thank you @dataflake and @d-maurer - I will dig into this topic again at this/next week's sprint and hopefully finish Chapter 4 of the Zope Dev book.

**Comment by dataflake** [comment]
You can probably understand the flow of action pretty well when you look at the code of `ZPublisher.WSGIPublisher.transaction_pubevents`, especially the code inside the `except`, which executes when publishing the resource has failed. I linked the code at the bottom.

First the exception is "translated" if possible to one of the exceptions from the `zExceptions` module in the call to `upgradeException`. That's what Dieter refers to in his last comment.

Then the header `x-wsgiorg.throw_errors` is looked up. If it exists, all further handling is circumvented and the exception is re-raised so it travels into the WSGI stack. This handling is under discussion in PR #294 right now, as that specific header apparently never became part of the WSGI standard.

The call to `_exc_view_created_response` attempts to look up and call a view for the exception. That function also contains code to support the old `standard_error_message` as fallback if there is no specific view registered for the exception. The `try`/`except` around it makes sure a failing exception view won't blow up further exception handling.

Next, if the exception is zException.Unauthorized it is re-raised so the browser can present a basic authentication dialog.

Next the exception is checked if it's one of the transient errors like a ZODB conflict error that can be re-tried for a given number of times before giving up.

The last step is a catch-all that basically says...

- if there's no exception view we can find or
- if this is `zExceptions.Unauthorized` or
- if the response object has the flag `debug_exceptions` set (see PR #563) 

... then the error is re-raised and will travel up the WSGI stack.

What happens when the exception is allowed into the WSGI stack depends entirely on the WSGI middleware or server you use. The default `waitress` reaction is that short error message. You can make `waitress` show a full traceback with its `expose_tracebacks` setting in its `[server]` definition inside the WSGI configuration file. Or you can hook in fancy debuggers like the `werkzeug` debugger. I wrote a shim egg that makes that easy, see https://dataflakewsgiwerkzeug.readthedocs.io/

https://github.com/zopefoundation/Zope/blob/42a2b0634fc656193c76d631b0881aebcf02537b/src/ZPublisher/WSGIPublisher.py#L162-L210

**Comment by dataflake** [comment]
P.S.: Not all of the logic inside `transaction_pubevents` is ideal and clean. I just documented what's there right now. There's always room for improvement.

**Comment by d-maurer** [comment]
Jens Vagelpohl wrote at 2019-5-5 18:20 -0700:
> ... steps for failure handling ...

Thank your for this excellent description of the failure handling steps!

>Next the exception is checked if it's one of the transient errors like a ZODB conflict error that can be re-tried for a given number of times before giving up.

I think this should be one of the first steps: "transient error"s
are not "real" errors - but a (likely resolvable) side effect
of unlucky circumstances coming together.
There is not much use to generate special views, translate messages
etc. if the request is retried.

**Comment by jugmac00** [comment]
see pr #596

**Comment by dataflake** [comment]
Fixed with the merge of #596
